dhcp-protocol/src/main/java/org/anarres/dhcp/v6/Dhcp6Exception.java

```diff
@@ -0,0 +1,239 @@
+package org.anarres.dhcp.v6;
+
+import com.google.common.annotations.Beta;
+import com.google.common.base.Preconditions;
+import java.util.Arrays;
+import javax.annotation.Nonnull;
+import org.anarres.dhcp.v6.messages.Dhcp6Message;
+import org.anarres.dhcp.v6.messages.Dhcp6MessageType;
+import org.anarres.dhcp.v6.options.ClientIdOption;
+import org.anarres.dhcp.v6.options.Dhcp6Option;
+import org.anarres.dhcp.v6.options.DuidOption;
+import org.anarres.dhcp.v6.options.IaNaOption;
+import org.anarres.dhcp.v6.options.IaTaOption;
+import org.anarres.dhcp.v6.options.RelayMessageOption;
+import org.anarres.dhcp.v6.options.ServerIdOption;
+import org.apache.directory.server.dhcp.DhcpException;
+
+/**
+ * Created by marosmars
+ */
+@Beta
+public class Dhcp6Exception extends DhcpException {
+
+    public Dhcp6Exception() {
+        super();
+    }
+
+    public Dhcp6Exception(@Nonnull final String description) {
+        super(description);
+    }
+
+    public Dhcp6Exception(@Nonnull final String description, @Nonnull final Exception e) {
+        super(description, e);
+    }
+
+    public static final class UnknownMsgException extends Dhcp6Exception {
+
+        private byte msgType;
+
+        public UnknownMsgException(byte msgType) {
+            super("Unrecognised msg type " + msgType);
+            this.msgType = msgType;
+        }
+
+        public byte getMsgType() {
+            return msgType;
+        }
+
+        /**
+         *
+         * Check if the message type is equal to DHCP_UNRECOGNIZED
+         *
+         * @throws UnknownMsgException in case the message is of unknown type
+         */
+        public static void check(@Nonnull final Dhcp6MessageType dhcp6MessageType, final byte type) throws UnknownMsgException {
+            if(dhcp6MessageType == Dhcp6MessageType.DHCP_UNRECOGNIZED) {
+                throw new UnknownMsgException(type);
+            }
+        }
+    }
+
+    public static final class InvalidMsgException extends Dhcp6Exception {
```

Comment:
Note to self: For every case where one of these is thrown, who do we expect to catch it, and does that cause sensible behaviour both from the PoV of the client and for our own internal state machine?